lib/logger/logger.go

```diff
@@ -130,16 +129,60 @@ func logLevelSkipframes(skipframes int, level, format string, args ...interface{
 func errorsAndWarnsLoggedCleaner() {
 	for {
 		time.Sleep(time.Second)
-		atomic.StoreUint64(&errorsLogged, 0)
-		atomic.StoreUint64(&warnsLogged, 0)
+		logLimiter.reset()
 	}
 }
 
 var (
-	errorsLogged uint64
-	warnsLogged  uint64
+	logLimiter = newLogLimit()
 )
 
+func newLogLimit() logLimit {
+	return logLimit{
+		s:  map[string]uint64{},
+		mu: sync.Mutex{},
+	}
+}
+
+type logLimit struct {
+	mu sync.Mutex
+	s  map[string]uint64
+}
+
+func (ll *logLimit) reset() {
+	ll.mu.Lock()
+	ll.s = make(map[string]uint64)
+	ll.mu.Unlock()
+}
+
+// needSuppress check if given string exceeds limit,
+// when count equals limit, log message prefix returned.
+func (ll *logLimit) needSuppress(file string, limit uint64) (bool, string) {
+	// fast path
+	var msg string
+	if limit == 0 {
+		return false, msg
+	}
+	ll.mu.Lock()
+	defer ll.mu.Unlock()
+
+	if n, ok := ll.s[file]; ok {
+		if n >= limit {
+			switch n {
+			// report only once
+			case limit:
+				msg = "suppressing log message with rate limit: "
```

Comment:
It would be great from user's PoV to see the limit value in log message. I.e. `msg = fmt.Sprintf("suppressing log message with rate limit=%d: ", limit)`